File: 2021/c/2/part1.c
#include <stdio.h>
#define MAX_CMD_SIZE 255

int main(int argc, char** argv)
{
        FILE *f;
        char cmd[MAX_CMD_SIZE];
        int err;
        int mag;
        int x;
        int y;

        f = fopen("input.txt", "r");
        if (!f) {
                return 1;
        }

        err = 0;
        x = 0;
        y = 0;
        while (2 == fscanf(f, "%s%d", cmd, &mag)) {
                printf("%s %d\n", cmd, mag);
                switch(cmd[0]) {
                        case 'f':
                                x += mag;
                                break;
                        case 'u':
                                y -= mag;
                                break;
                        case 'd':
                                y += mag;
                                break;
                        default:
                                printf("unknown command\n");
                                err = 1;
                                goto cleanup;

                }
                printf("x: %d\ty: %d\n", x, y);
        }
        printf("%d\n", x * y);

cleanup:
        fclose(f);
        return err;
}

File: 2021/rs/2/part1.rs
use std::fs;
use std::io;
use std::io::prelude::*;

fn main() -> io::Result<()> {
    let file = fs::File::open("input.txt")?;
    let br = io::BufReader::new(file);

    let mut x: i32 = 0;
    let mut y: i32 = 0;
    for line in br.lines() {
        let unwrapped = line.expect("Failed to read line.");
        let mut split = unwrapped.split(" ");
        let cmd = split.next().expect("Failed to get command token.");
        let mag: i32 = split
            .next()
            .expect("Failed to get magnitude token.")
            .parse()
            .expect("Failed to parse magnitude.");
        match cmd {
            "forward" => x += mag,
            "up" => y -= mag,
            "down" => y += mag,
            _ => panic!("Unknown command"),
        }
    }

    println!("x: {}\ty: {}\n{}", x, y, x * y);

    Ok(())
}
